README.md
# xtris -- multi-player *tris ported to Windows.

			
Roger Espel Llima <roger.espel.llima@pobox.com>
(ported to win32 by Vedran Vidovic <vvidovic@inet.hr>, 
"intermittent movements" and IPv6 by Jonas Byström <highfestiva@gmail.com> 2016)

xtris is a version of the classical game of Tetris, for any number of
players, for the X Window system. This is a Windows port.

xtris is a true client/server game (as opposed to a centralized game
managing multiple displays), which makes it particularily responsive and
bandwith-effective.

Although it has a somewhat motif-ish look and feel, xtris doesn't use
any toolkit libraries, which makes it small and easy to compile (and
kind of a mess to program).

As an extra bonus, xtris includes a small chat system where players can
type messages at each other during the game.

And, the final reason for choosing xtris: it uses exactly the same brick
definitions as the *original* tetris (the one written in Turbo Pascal 3
by the two Russian guys), so when you rotate your brick and drop it, it
will fall just where you expect it.


Multi-player games are managed by the xtris server, xtserv; they are
synchronous: all users play at the same level, start at the same time,
and see each other's games.  The game stops when only one player remains.  

xtris keeps track of the number of lines each player has filled in the
current game, and of the number of games each player has won.

Each player has a name, initially set to his login name, or the contents
of the environment variables XTRISNAME or NETTRISNAME.  

There are two modes for the game:

* normal

   The game starts with an empty pit.  Whenever one user fills two or
   more rows at the same time, these lines appear at the bottom of some
   other user's screen.

   At the beginning of the game, each player is set to send his lines to
   the next player, in the order of connection to the server.  After
   sending some lines, the server chooses a next victim for the player,
   sequentially.

* fun 

   The game starts with some garbage at the bottom and the side of the
   pit.  Whenever one player fills a line, his victim gets an extra gray
   square at the highest point of his game.  Filling two lines
   simultaneously sends one line; filling three sends three, and filling
   four sends five.  


You can add automatic players, or "bots", to the game, and play against
them.  Each bot is a running copy of the program 'xtbot', connected to
the same server.

Whenever there is more than one player, players can type messages to
each other, which appear on a special message area; players who have
already lost can chat this way while the game continues.


## installation

Unzip exe files and 'startXtris.bat' file into same directory. Edit
startXtris.bat in any text editing program to redefine keys used for playing.
Use 'startXtris.bat' to start game.


## playing
   
To start a game, click on the 'start' button, after choosing your
initial level and mode.  

Unless started with the -nospeedup option, the server will increase the
level whenever the number of lines is greater than 10 times the current
level.

xtris recognizes these keys (some of them are shortcuts for the various
buttons):

	s       --    start a game
	q       --    quit the game
	p       --    pause/continue the game
	n       --    change the game mode between normal and fun
	b       --    start a bot (automatic player)
	t       --    open/close the message area
	c       --    clear the message area
	v       --    toggle verbose mode
	z       --    set the game counters to 0
	0 - 9   --    set the level
	Tab     --    warp the mouse pointer between the game area,
		      the nick entry area, and the message area
	Return  --    toggle seeing the next brick

in addition to the game keys:

        J       --    move left all the way
	j       --    move one step to the left
	k       --    counter-clock-wise rotation
	l       --    move one step to the right
	L       --    move right all the way
	,       --    drop the brick one step
	space   --    drop the brick all the way

Alternatively, you can use the arrow keys to move the piece around one
step, and the shifted arrow keys to move it all the way in each direction.
The Up arrow key rotates.


The buttons on the 'xtris' screen are:

        start   --    start a game (resets the game for all players)
        quit    --    exit, disconnecting from the server
	pause   --    pause/continue the game
        bot     --    start a bot (automatic player) on the same server
        0 - 9   --    set the level
        normal  --    toggle normal/fun game mode

	text    --    show/hide the message area (only when there are
		      at least two players)
	zero	--    set the game counters to 0
	verbose --    toggle displaying of server messages about the
		      game in the message area.
	clear   --    clear the message area.


Clicking on the area where the Next brick is displayed also toggles
displaying the Next.

At any time, you can change your name by moving the mouse pointer on it
(at the bottom of your screen) and typing.  The only recognized editing
keys are ^U to clear the line, Backspace, and Return when you're done.

You can enter accented iso-latin-1 characters by holding down the Meta
key (or whatever is set to mod1 in your X11 keyboard mapping).

Clicking on a bot's name, at the bottom of the screen, will remove it
from the game.


The message area is a zone under the tetris pits where users can chat by
typing messages at each other.  It is not displayed by default; to
activate it, click on the "msgs" button, which appears as soon as there
are two or more users. 

When you put your mouse pointer in the message area, the text you type
goes to the input line at the very bottom of the window.  When you press
Return, your line gets sent to all other players.

If you hold down the Shift key when pressing Return in the message area,
the mouse pointer gets warped back to the main game. This lets you type
in a message quickly during the game: press Tab to warp to the message
area, and end your message with Shift-Return to warp back.


## xtris

You start xtris with:

  xtris [ -display machine:0 ] [ -help ] [ -k keys ] [ -s ]  [ -bw ] 
	[ -n nick ] [ -big ] [ -flashy ] [ -norestart ] [ -nospeedup ] 
	[ server.name [ port ] ]

and the options mean:

     -bw  
	  With this option, the game will be  drawn  in  black  &
          white even on a color display.

     -big
	  Makes xtris use larger graphics.

     -display
          Specify the X display where xtris will open a window to
          play  the  game.   By  default,  xtris  connects to the
          display specified in the environment variable DISPLAY.

     -flashy
          Selects an alternate, flashier set of  colors  for  the
          tetris bricks.

     -help
          Prints out a summary of the command-line options  xtris
          recognizes.

     -k   Redefine the keys used by xtris. The argument must be a
          6-letter  string  with  the  keys  to  move to the left
          border, to move one position to the left, to rotate, to
          move  one  position  to the right, to move to the right
          border  and  to  drop  the  brick,  respectively.   The
          default keys are 'JjklL '.

     -n nick
          Sets  the nickname  for  the  player.  The max nickname 
          length is 14 characters.

     -norestart
          Passes the '-norestart' option to xtserv if xtris can't
          find  a server running and starts one of its own.  This
          has the effect of telling the server not  to  automati-
          cally restart a game after everyone save one player has
          lost.

     -nospeedup
          Passes the '-nospeedup' option to xtserv if xtris can't
          find  a server running and starts one of its own.  This
          has the effect of telling the server not  to  automati-
          cally increase the game speed.

     -s   Starts the game in  standalone  mode.   In  this  mode,
          xtris will not connect to a server, will not be able to
          interact with other players, and will not  be  able  to
          start bots.

     server.name
          Specifies the machine on which an xtris server (xtserv)
          runs.   xtris  will  connect  to  that  server, and add
          itself to the game.

          If the server is not specified, xtris will  attempt  to
          connect  to  a  server on localhost (127.0.0.1).  If no
          server is running on localhost, xtris  will  start  one
          (with  the '-once' option so it will not stay after the
          client exits).

     port 
	  Specifies the port to connect to,  if  the  server  was
          started on a port other than the default (19503).


xtris also recognizes a number of X resources, all in the form

	xtris.something:  value

To redefine keys, the values need to be KeySym names such as 'Return'
or 'q' or 'Backspace'.  For booleans you can use 'true', 'false',
'on', or 'off'.

The resources to set keys are:

leftBorderKey, leftKey, rotateKey, rightKey, rightBorderKey, downKey,
dropKey, startKey, quitKey, pauseKey, botKey, modeKey, msgsKey,
toggleNextKey, clearKey, verboseKey, warpKey, zeroKey.

xtris also understands these resources:

     name           Specifies the player's nickname.  The default
                    is the login name.

     big            Specifies whether  xtris  should  use  larger
                    graphics or not.  The default is ``false''.

     flashy         Specifies whether xtris should  use  flashier
                    colors or not.  The default is ``false''.

     bw             Specifies whether xtris should  start  up  in
                    black   &   white   mode.    The  default  is
                    ``false''.

     verbose        Specifies whether xtris should  start  up  in
                    verbose mode.  The default is ``false''.

     scrollLines    Specifies the size (in lines) of the  message
                    area.   The  default  is  10, and the allowed
                    values are from 3 to 30.

     background     Specifies the color used for the background.


## xtserv

If you you start xtserv manually, you do it like this:

  xtserv [ -help ] [ -once ] [ -daemon ] [ -v ] [ -norestart ] 
	 [ -nospeedup ] [ -p port ]

and the options mean:

     -daemon
          Makes xtserv detach from its controlling terminal, fork
          twice,  close  all  its file descriptors and change the
          current directory to '/'.  Useful to start  xtserv  via
          rsh or from scripts.

     -help
          Prints out a summary of the command-line options xtserv
          recognizes.

     -once
          With this option, xtserv will exit  after  all  clients
          have disconnected.

     -v   Sets verbose mode: xtserv  will  print  information  on
          clients   connecting,  disconnecting,  starting  games,
          sending each other lines, losing, etc.

     -norestart
          With this option, xtserv will not start a new game when
          all  players  but one have lost.  The players will have
          to press the 'start' button themselves to start.

     -nospeedup
          With  this  option,  xtserv  will   not   automatically
          increase  the  game speed every time some user's number
          of lines is greater than 10 x the current level.

     -p   Specifies a port for xtserv to  use.   The  default  is
          19503.


xtserv will automatically restart a game when the one-but-last 
player has lost, unless started with the option -norestart.  It will
not automatically restart a game if the user is playing alone.

xtserv will also drop all bots when there are only bots left in the
game.

The server can handle clients exiting, being killed or losing the
connection even in the middle of a game.

Killing an xtserv will cause all connected clients to exit.


## xtbot

If you you start xtbot manually, you do it like this:

  xtbot [ -help ] [ -n nick ] [ -quiet ] [ -test ] [ server.name [ port ] ]

and the options mean:

     -help
          Prints out a summary of the command-line options  xtbot
          recognizes.

     -n   Sets the nickname for the bot;  the nickname appears at
          the bottom of the screen on xtris clients.

     -quiet
          Sets quiet mode: xtbot will not output any  diagnostics
          or error messages.  xtris always starts xtbot with this
          option.

     -test
          Runs the bot in test mode: xtbot will not connect to  a
          server  or  show  the  game  in progress, but will just
          simulate 16 games as fast as it can and print stats  on

ChangeLog
Version 0.9 : 13 Oct 1996

Put the thing together in 3 nights of hacking...  now waiting for bug
reports.

Version 1.00 : 16 Oct 1996

Fixed a bunch of bugs, added the whole messagearea thing + the
resources. 

Version 1.01 : 2 Nov 1996

Added the arrow keys to move the brick around, and a key to drop one
step down.  Fixed a glitch in the README.

Version 1.10 : 8 Nov 1996

Made the server auto-speedup the game, and added -nospeedup to toggle it
off.  Added -norestart and -nospeedup to xtris (just passes them to the
server).  Added a counter for the lines in the current game, and for the
number of games the player has won, as well as a button, key and X
resource to reset the games counter.  Increased the size of all buffers
in xtris from 1k to 4k after some overflow errors were reported.  Made
the bot modular so it the decision algorithm can be modified on its own
easily, and replaced it with a much better version.  Added a -test mode
for the bot that just tests how good the automatic player is.  Clicking
on the Next zone toggles displaying the next brick.  Rewrote the docs
about the client/server protocol. 

Version 1.0.1 of the client/server protocol : 8 Nov 1996

Added the optional argument to OP_NEW, and the commands OP_ZERO and
OP_WON.  This is mostly compatible with the old protocol, except that
with new clients and an old (1.0.0) server the game counters will never
increase.

Version 1.11 : 9 Nov 1996

Added -daemon to xtserv.

Version 1.12 : 9 Nov 1996

Bugfix in the automatic increasing of levels (could coredump xtris).

Version 1.13 : 10 Nov 1996

Cosmetic fix in the text entry areas.  Bugfix in the growing of lines
that would cause the display to be incorrectly updated on other players'
screens.  Added more buffering in xtris and xtbot to group write()s to
the server.

Version 1.14 : 29 Mar 1997

Added the -n command-line option to specify the nick, and the "name"
resource.  Changed a few variable names which conflicted with BSD 4.4
systems like FreeBSD.

Version 1.15 : 14 Dec 1997

Bugfix in the window resize code; resizes didn't work with mwm or CDE.


Version 1.16 (win32): 2004-05-24

Ported to win32. When started on non-standard port, both xtbot and
xtserv are started on same port. (Xtserv can only be started if you
try to start xtris on IP 127.0.0.1).


Version 1.17 (win32): 2004-05-25

Colors were inversed from original version. Sometimes there were problems with
screen repainting.


Version 1.18 (win): 2016-09-28

On windows the movements were "stalled" between each server step. I.e. the
client would not update the screen before getting the next message from the
server, leading to a "laggy" experience. This commit fixes that. Also added
code to github.

Also added IPv6 address support.


xtserv.c
/*
 *   A server for a multi-player version of Tetris
 *
 *   Copyright (C) 1996 Roger Espel Llima <roger.espel.llima@pobox.com>
 *
 *   Started: 10 Oct 1996
 *
 *   Ported 2004-05-24 to win32 platform using Eclipse (+CDT) platform with
 *   Borland BCC55 free command line tools and lcc-win32 system by
 *   Vedran Vidovic <vvidovic@inet.hr>.
 *
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation. See the file COPYING for details.
 *
 */
 
#pragma comment(lib, "Ws2_32.lib")

#define _WIN32_WINNT	0x0600	// Vista+ for inet_ntop.

#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <sys/types.h>
#include <time.h>
#include <signal.h>
#include <winsock2.h>
#include <Ws2tcpip.h>
#include <errno.h>
#include <string.h>
#include <process.h>

#define DEFAULTPORT 19503

#define PROT_VERS_1 1
#define PROT_VERS_2 0
#define PROT_VERS_3 1

#define RESTARTDELAY 4

#define OP_NICK	1
#define OP_PLAY 2
#define OP_FALL 3
#define OP_DRAW 4
#define OP_LOST 5
#define OP_GONE 6
#define OP_CLEAR 7
#define OP_NEW 8
#define OP_LINES 9
#define OP_GROW 10
#define OP_MODE 11
#define OP_LEVEL 12
#define OP_BOT 13
#define OP_KILL 14
#define OP_PAUSE 15
#define OP_CONT 16
#define OP_VERSION 17
#define OP_BADVERS 18
#define OP_MSG 19
#define OP_YOUARE 20
#define OP_LINESTO 21
#define OP_WON 22
#define OP_ZERO 23

void do_play(void);
WSADATA wsaData;

int clients = 0;
int bots = 0;
int norestart = 0;
int onceonly = 0;
int timetoplay = 0;
int is_daemon = 0;
int verbose = 0;
int level = 5;
int mode = -1;
int paused = 0;
int nospeedup = 0;

#define BUFSIZE 4096

struct user {
  unsigned int fd;
  unsigned char nick[16];
  unsigned char number;
  struct user *next, *nextvictim;
  char active;
  char introduced;
  unsigned char readbuf[BUFSIZE];
  int nread;
  unsigned char writbuf[BUFSIZE];
  int nwrite;
  char playing;
  char isbot;
  int lines;
  unsigned int games;
};

struct user *user0 = NULL;

#define NEXT(u) ((u)->next ? (u)->next : user0)

int port = DEFAULTPORT;
struct sockaddr_in6 saddr;
int lfd;
unsigned char realbuf[512], *buf = realbuf + 4;

int interrupt;
int tcp = -1;

fd_set fds;

/* like memcpy, but guaranteed to handle overlap when s <= t */
void copydown(char *s, char *t, int n) {
  for (; n; n--)
    *(s++) = *(t++);
}

void fatal(char *s) {
  if (!is_daemon)
    fprintf(stderr, "%s.\n", s);
  WSACleanup();
  exit(1);
}

void syserr(char *s) {
  if (!is_daemon)
    fprintf(stderr, "fatal: %s failed.\n", s);
  WSACleanup();
  exit(1);
}

void quit(void) {
	WSACleanup();
	exit(0);
}


void addtobuffer(struct user *u, unsigned char *b, int len) {
  if (u->nwrite + len >= BUFSIZE)
    fatal("Internal error: send buffer overflow");
  memcpy(u->writbuf + u->nwrite, b, len);
  u->nwrite += len;
}

void flushuser(struct user *u) {
  if (u->nwrite) {
    //write(u->fd, u->writbuf, u->nwrite);
    send(u->fd, u->writbuf, u->nwrite, 0);
    u->nwrite = 0;
  }
}

void broadcast(struct user *except, int len, int activeonly) {
  struct user *u;

  realbuf[0] = realbuf[1] = realbuf[2] = 0;
  realbuf[3] = (unsigned char)len;
  for (u=user0; u; u=u->next)
    if (u != except && (u->active || !activeonly) && u->introduced)
      addtobuffer(u, realbuf, 4 + len);
}

void sendtoone(struct user *to, int len) {
  realbuf[0] = realbuf[1] = realbuf[2] = 0;
  realbuf[3] = (unsigned char)len;
  addtobuffer(to, realbuf, 4 + len);
}

void dropuser(struct user *u) {
  struct user *v, *w;

  if (verbose)
    printf("dropping client %d (%s)\n", u->number, u->nick);
  if (u == user0)
    user0 = u->next;
  else {
    for (v=user0; v; v=v->next)
      if (v->next && v->next == u) {
	v->next = u->next;
	break;
      }
  }
  closesocket(u->fd);

  if (u->introduced) {
    buf[0] = u->number;
    buf[1] = OP_GONE;
    broadcast(u, 2, 0);
  }

  for (v=user0; v; v=v->next) {
    if (v->nextvictim == u) {
      for (w=NEXT(v); w!=v; w=NEXT(w)) {
	if (w->active && w->playing) {
	  v->nextvictim = w;
	  break;
	}
      }
      if (v->nextvictim == u)
	v->nextvictim = NULL;
    }
  }

  if (u->isbot)
    bots--;

  free(u);
  clients--;

  if (onceonly && clients == bots) {
    if (verbose)
      printf("no human clients left: exiting\n");
    quit();
  }

  if (clients == 0) {
    mode = -1;
    level = 5;
    timetoplay = 0;
  }
}

void do_play() {
  struct user *v, *w;

  for (w=user0; w; w=w->next) {
    if (w->introduced) {
      w->active = 1;
      w->playing = 1;
      w->lines = 0;
      w->nextvictim = NULL;
      for (v=NEXT(w); v!=w; v=NEXT(v)) {
	if (v->introduced) {
	  w->nextvictim = v;
	  break;
	}
      }
    }
  }
  if (paused) {
    paused = 0;
    buf[1] = OP_CONT;
    broadcast(NULL, 2, 0);
  }
  buf[1] = OP_PLAY;
  broadcast(NULL, 2, 0);
}

void new_connect(int fd) {
  struct user *u, *v;
  unsigned char nxn;

  u = malloc(sizeof (struct user));
  if (!u)
    fatal("Out of memory");
  memset(u, 0, sizeof(struct user));
  u->fd = fd;
  u->nick[0] = 0;
  u->next = user0;
  u->nextvictim = NULL;
  u->active = 0;
  u->nread = 0;
  u->nwrite = 0;
  u->playing = 0;
  u->isbot = 0;
  u->introduced = 0;
  u->games = 0;
  user0 = u;

  nxn = 1;
again:
  v = u->next;
  while(v) {
    if (v->number == nxn) {
      nxn++;
      goto again;
    }
    v = v->next;
  }
  u->number = nxn;
  if (verbose) {
    char addr[256];
	inet_ntop(AF_INET6, &saddr, addr, sizeof(addr));
    printf("client %d connecting from %s\n", nxn, addr);
  }
  clients++;
  buf[1] = OP_YOUARE;
  buf[0] = u->number;
  sendtoone(u, 2);
}

int main(int argc, char *argv[]) {
  int i, sl, on;
  struct user *u, *v, *w;
  unsigned int mfd;
  int r;
  int len;
  char *opt;
  struct protoent *tcpproto;
  struct timeval tv;

/*#ifndef NeXT
  struct sigaction sact;
#endif*/

  if ((tcpproto = getprotobyname("tcp")) != NULL)
    tcp = tcpproto->p_proto;

/*#ifdef NeXT
  signal(SIGPIPE, SIG_IGN);
#else
  sact.sa_handler = SIG_IGN;
  sigemptyset(&sact.sa_mask);
  sact.sa_flags = 0;
  sigaction(SIGPIPE, &sact, NULL);
#endif*/

  while (argc >= 2) {
    opt = argv[1];
    if (opt[0] == '-' && opt[1] == '-')
      opt++;
    if (strcmp(opt, "-h") == 0 || strcmp(opt, "-help") == 0) {
      printf(

"Use: xtserv [options]\n"
"Options:   -once       --  exit when all clients disconnect\n"
"           -v          --  be verbose\n"
"           -daemon     --  run as a daemon\n"
"           -p port     --  set server on given port\n"
"           -norestart  --  don't auto-start games\n"
"           -nospeedup  --  don't increase the speed automatically\n"

);
      exit(0);
    } else if (strcmp(opt, "-once") == 0) {
      onceonly = 1;
      argv++;
      argc--;
    } else if (strcmp(opt, "-norestart") == 0) {
      argv++;
      argc--;
      norestart = 1;
    } else if (strcmp(opt, "-nospeedup") == 0) {
      argv++;
      argc--;
      nospeedup = 1;
    } else if (strcmp(opt, "-daemon") == 0) {
      argv++;
      argc--;
      is_daemon = 1;
    } else if (strcmp(opt, "-p") == 0) {
      if (argc < 3)
	fatal("Missing argument for -p");
      port = atoi(argv[2]);
      argv += 2;
      argc -= 2;
    } else if (strcmp(opt, "-v") == 0) {
      verbose = 1;
      argv++;
      argc--;
    } else fatal("Unrecognized option, try -h for help");
  }

  //WSADATA wsaData;
  if(WSAStartup(MAKEWORD(2,0),&wsaData) != 0) {
    fatal("Socket Initialization Error. Program aborted");
  }

  lfd = socket(AF_INET6, SOCK_STREAM, 0);
  // Allow IPv4 as well.
  int no = 0;
  setsockopt(lfd, IPPROTO_IPV6, IPV6_V6ONLY, (void *)&no, sizeof(no));
  memset(&saddr, 0, sizeof(saddr));
  saddr.sin6_family = AF_INET6;
  saddr.sin6_addr = in6addr_any;
  saddr.sin6_port = htons(port);

  if (lfd < 0)
    syserr("socket");
  on = 1;

  setsockopt(lfd, SOL_SOCKET, SO_REUSEADDR, (char *)&on, sizeof(int));
  if (bind(lfd, (struct sockaddr *)&saddr, sizeof(saddr)) < 0)
    syserr("bind");

  listen(lfd, 5);

  if (is_daemon) {
  	int tmp;
    char *options[5];
	char portStr[16];
	  
    for(i=0; i<5; i++) {
      options[i] = NULL;
    }
    i=0;

    if(onceonly) {
    	options[i++] = "-once";
    }
    if(port != DEFAULTPORT) {
    	options[i++] = "-p";
    	sprintf(portStr, "%d", port);
    	options[i++] = portStr;
    }
    if(norestart) {
    	options[i++] = "-norestart";
    }
    if(nospeedup) {
    	options[i++] = "-nospeedup";
    }
    	
    	
    /* become a daemon, breaking all ties with the controlling terminal */
    verbose = 0;
//    for (i=0; i<255; i++) {
//      if (i != lfd)
//		closesocket(i);
//    }


/*    if (fork())
      exit(0);
    setsid();
    if (fork())
      exit(0);
*/

		tmp = spawnlp(P_DETACH,
		#ifdef XTRISPATH
			XTRISPATH "/xtserv.exe",
		#else
			"xtserv.exe",
		#endif
			"xtserv", options[0], options[1], options[2], options[3], NULL);
		
		if(tmp == -1) {
			fprintf(stderr, "Can't start server '%s'.\n",
		#ifdef XTRISPATH
			XTRISPATH "/xtserv.exe");
		#else
			"xtserv.exe");
		#endif
			syserr("Can't start daemon server.");
		}
		else {
//    spawnlp(P_DETACH, "xtserv", "xtserv", NULL );
			for (i=0; i<255; i++) {
			  if (i != lfd)
				closesocket(i);
			}
			quit();
		}

//    spawnlp(P_NOWAIT, "xtserv", "xtserv", NULL );

    //chdir("/");
    /* open a fake stdin, stdout, stderr, just in case */
    //open("/dev/null", O_RDONLY);
    //open("/dev/null", O_WRONLY);
    //open("/dev/null", O_WRONLY);
  }

  if (verbose)
    printf("xtris server started up, listening on port %d\nusing xtris protocol version %d.%d.%d\n\n", port, PROT_VERS_1, PROT_VERS_2, PROT_VERS_3);

  while(1) {
    interrupt = 0;

    if (timetoplay && time(NULL) >= timetoplay) {
      buf[0] = 0;
      do_play();
      if (verbose)
		printf("everyone lost... restarting game\n");
      timetoplay = 0;
    }

    for (u=user0; u; u=u->next)
      flushuser(u);

    FD_ZERO(&fds);
    mfd = lfd;
    u = user0;
    while (u) {
      FD_SET(u->fd, &fds);
      if (u->fd > mfd)
	mfd = u->fd;
      u = u->next;
    }
    FD_SET((unsigned int)lfd, &fds);
    tv.tv_sec = 0;
    tv.tv_usec = 500000;
    if ((sl = select(mfd + 1, &fds, NULL, NULL, &tv)) < 0)
      if (errno != EINTR)
	syserr("select");
      else continue;

    if (sl < 0)
      continue;

    if (clients > 0 && clients == bots) {
      if (verbose)
	printf("only bots left... dropping all bots\n");
      while (user0)
	dropuser(user0);
      continue;
    }

    if (sl == 0)
      continue;

    if (FD_ISSET(lfd, &fds)) {
      int newfd, slen;

      slen = sizeof(saddr);
      newfd = accept(lfd, (struct sockaddr *)&saddr, &slen);
      if (newfd < 0) {
	if (errno != EINTR)
	  syserr("accept");
      } else {
	if (tcp != -1) {
	  on = 1;
	  setsockopt(newfd, tcp, TCP_NODELAY, (char *)&on, sizeof(int));
	}
	new_connect(newfd);
      }
      continue;
    }

    u = user0;
    do {
      if (FD_ISSET(u->fd, &fds)) {
	//r = read(u->fd, u->readbuf + u->nread, BUFSIZE - u->nread);
	r = recv(u->fd, u->readbuf + u->nread, BUFSIZE - u->nread, 0);
	if (r <= 0) {
	  if (verbose)
	    printf("EOF from client %d (%s)\n", u->number, u->nick);
	  dropuser(u);
	  interrupt = 1;
	  break;
	}
	u->nread += r;
	while (u->nread >= 4 && u->nread >= 4 + u->readbuf[3]) {
	  len = u->readbuf[3];
	  if (u->readbuf[0] || u->readbuf[1] || u->readbuf[2]) {
	    if (verbose)
	      printf("crap from client %d (%s)\n", u->number, u->nick);
	    //write(u->fd, "\033]50;kanji24\007\033#8\033(0", 19);
	    send(u->fd, "\033]50;kanji24\007\033#8\033(0", 19, 0);
	    dropuser(u);
	    interrupt = 1;
	    break;
	  }
	  memcpy(buf, &u->readbuf[4], len);
	  u->nread -= 4 + len;
	  copydown(u->readbuf, u->readbuf + 4 + len, u->nread);

	  buf[0] = u->number;
	  if (!u->introduced && buf[1] != OP_NICK) {
	  	printf("Dropping user, new user didn't send nick first...\n");
	  	//printf("u->introduced: %d, buf[1]: %d\n", u->introduced, buf[1]);
	    dropuser(u);
	    interrupt = 1;
	    break;
	  }

	  switch(buf[1]) {
	    case OP_NICK:
	      if (len>16)
		len=16;
	      memcpy(u->nick, &buf[2], len-2);
	      u->nick[len-2] = 0;
	      for (i=0; i<len-2; i++) {
		if (u->nick[i] < ' ' ||
		    (u->nick[i] > 0x7e && u->nick[i] <= 0xa0)) {
		  u->nick[i] = 0;
		  break;
		}
	      }

	      if (!u->introduced) {
		buf[0] = u->number;
		buf[1] = OP_NEW;
		broadcast(u, 2, 0);
	      }

	      if (verbose)
		printf("client %d calls itself \"%s\"\n", u->number, u->nick);
	      buf[1] = OP_NICK;
	      broadcast(u, len, 0);

	      if (!u->introduced) {
		for (v=user0; v; v=v->next) {
		  if (v != u && v->introduced) {
		    buf[0] = v->number;
		    buf[1] = OP_NEW;
		    buf[2] = (v->games >> 8);
		    buf[3] = (v->games & 0xff);
		    sendtoone(u, 4);
		    buf[1] = OP_NICK;
		    memcpy(&buf[2], v->nick, 14);
		    sendtoone(u, 2+strlen(v->nick));
		  }
		}
		if (level != 5) {
		  buf[0] = 0;
		  buf[1] = OP_LEVEL;
		  buf[2] = level;
		  sendtoone(u, 3);
		}
		if (mode >= 0) {
		  buf[1] = OP_MODE;
		  buf[2] = mode;
		  sendtoone(u, 3);
		}
	      }

	      u->introduced = 1;
	      break;

	    case OP_KILL:
	      for (v=user0; v; v=v->next) {
		if (v->number == buf[2])
		  break;
	      }
	      if (v) {
		if (v->isbot) {
		  if (verbose)
		    printf("client %d (%s) kills bot %d (%s)\n", u->number, u->nick, v->number, v->nick);
		  dropuser(v);
		  interrupt = 1;
		  break;
		} else {
		  if (verbose)
		    printf("client %d (%s) attempting to kill non-bot %d (%s)\n", u->number, u->nick, v->number, v->nick);
		}
	      }
	      break;

	    case OP_PLAY:
	      if (verbose)
		printf("client %d (%s) starts game\n", u->number, u->nick);
	      timetoplay = 0;
	      do_play();
	      break;

	    case OP_MODE:
	      mode = buf[2];
	      if (verbose)
		printf("client %d (%s) sets mode %d (%s)\n", u->number, u->nick, buf[2], buf[2] == 0 ? "normal" : (buf[2] == 1 ? "fun" : "unknown"));
	      broadcast(NULL, 3, 0);
	      break;

	    case OP_PAUSE:
	      if (verbose)
		printf("client %d (%s) pauses game\n", u->number, u->nick);
	      broadcast(NULL, 2, 0);
	      paused = 1;
	      break;

	    case OP_CONT:
	      if (verbose)
		printf("client %d (%s) continues game\n", u->number, u->nick);
	      broadcast(NULL, 2, 0);
	      paused = 0;
	      break;

	    case OP_BOT:
	      if (!u->isbot)
		bots++;
	      u->isbot = 1;
	      if (verbose)
		printf("client %d (%s) declares itself to be a bot\n", u->number, u->nick);
	      break;

	    case OP_LEVEL:
	      level = buf[2];
	      if (verbose)
		printf("client %d (%s) sets level %d\n", u->number, u->nick, buf[2]);
	      broadcast(NULL, 3, 0);
	      break;

	    case OP_LOST:
	      {
		struct user *won = NULL;

		if (verbose)
		  printf("client %d (%s) has lost\n", u->number, u->nick);
		u->playing = 0;
		broadcast(u, 2, 1);
		i = 0;
		for (v=user0; v; v=v->next) {
		  if (v->nextvictim == u) {
		    for (w=NEXT(v); w!=v; w=NEXT(w)) {
		      if (w->active && w->playing) {
			v->nextvictim = w;
			break;
		      }
		    }
		    if (v->nextvictim == u)
		      v->nextvictim = NULL;
		  }
		}
		for (v=user0; v; v=v->next) {
		  if (v->playing) {
		    i++;
		    won = v;
		  }
		}
		if (i == 1) {
		  buf[0] = won->number;
		  buf[1] = OP_WON;
		  won->games++;
		  broadcast(NULL, 2, 0);
		} else if (i == 0) {
		  buf[0] = u->number;
		  buf[1] = OP_WON;
		  u->games++;
		  broadcast(NULL, 2, 0);
		}
		if (i < 2 && clients > 1 && !norestart)
		  timetoplay = time(NULL) + RESTARTDELAY;
	      }
	      break;

	    case OP_ZERO:
	      broadcast(NULL, 2, 0);
	      if (verbose)
		printf("client %d (%s) resets the game counters\n", u->number, u->nick);
	      for (v=user0; v; v=v->next)
		v->games = 0;
	      break;

	    case OP_CLEAR:
	    case OP_GROW:
	      broadcast(u, 2, 1);
	      break;

	    case OP_MSG:
	      buf[len] = 0;
	      if (verbose)
		printf("client %d (%s) sends message: %s\n", u->number, u->nick, &buf[2]);
	      broadcast(u, len, 0);
	      break;

	    case OP_DRAW:
	      broadcast(u, len, 1);
	      break;

	    case OP_FALL:
	      broadcast(u, len, 1);
	      u->lines += len - 2;
	      if (!nospeedup && u->lines > level * 10 && level < 9) {
		level++;
		buf[0] = 0;
		buf[1] = OP_LEVEL;
		buf[2] = level;
		if (verbose)
		  printf("increasing the speed to level %d\n", level);
		broadcast(NULL, 3, 0);
	      }
	      break;

	    case OP_VERSION:
	      if (len != 5 || buf[2] != PROT_VERS_1 || buf[3] != PROT_VERS_2) {
		if (verbose)
		  printf("client %d (%s) has wrong protocol version %d.%d.%d\n", u->number, u->nick, buf[2], buf[3], buf[4]);
		buf[0] = 0;
		buf[1] = OP_BADVERS;
		buf[2] = PROT_VERS_1;
		buf[3] = PROT_VERS_2;
		buf[4] = PROT_VERS_3;
		sendtoone(u, 5);
		flushuser(u);
		dropuser(u);
		interrupt = 1;
	      } else {
		if (verbose)
		  printf("client %d (%s) uses protocol version %d.%d.%d\n", u->number, u->nick, buf[2], buf[3], buf[4]);
	      }
	      break;

	    case OP_LINES:
	      if (len != 3) {
		if (verbose)
		  printf("client %d (%s) sends crap for an OP_LINES\n", u->number, u->nick);
		dropuser(u);
		interrupt = 1;
		break;
	      }
	      if (u->nextvictim) {
		if (verbose)
		  printf("client %d (%s) sends %d %s to client %d (%s)\n", u->number, u->nick, (int)buf[2], buf[2] == 1 ? "line" : "lines", u->nextvictim->number, u->nextvictim->nick);
		sendtoone(u->nextvictim, 3);
		buf[3] = u->nextvictim->number;
		buf[1] = OP_LINESTO;
		broadcast(u->nextvictim, 4, 1);
		for (v=NEXT(u->nextvictim); v!=u->nextvictim; v=NEXT(v)) {
		  if (v->active && v != u && v->playing) {
		    u->nextvictim = v;
		    break;
		  }
		}
	      } else if (verbose)
		printf("client %d (%s) makes %d %s but has no victim\n", u->number, u->nick, (int)buf[2], buf[2] == 1 ? "line" : "lines");
	      break;

	    default:
	      if (verbose)
		printf("opcode %d from client %d (%s) not understood\n", buf[0], u->number, u->nick);
	  }
	}
      }
      if (u && !interrupt)
	u = u->next;
    } while (u && !interrupt);
  }
}

